gensim/models/doc2vec.py

```diff
@@ -789,14 +802,101 @@ def build_vocab_from_freq(self, word_freq, keep_raw_vocab=False, corpus_count=No
+    document_no = -1
+    total_words = 0
+    for document_no, document in enumerate(stream):
+        if not checked_string_types:
+            if isinstance(document.words, string_types):
+                log_msg = "Each 'words' should be a list of words (usually unicode strings). " \
+                          "First 'words' here is instead plain %s." % type(document.words)
+                progress_queue.put(log_msg)
+
+            checked_string_types += 1
+
+        document_length = len(document.words)
+
+        for tag in document.tags:
+            doclen2tags[document_length].append(tag)
+
+        for word in document.words:
+            vocab[word] += 1
+        total_words += len(document.words)
+
+        if max_vocab_size and len(vocab) > max_vocab_size:
+            utils.prune_vocab(vocab, min_reduce, trim_rule=trim_rule)
+            min_reduce += 1
+
+    progress_queue.put((total_words, document_no + 1))
+    progress_queue.put(None)
+    return vocab, doclen2tags
+
+
 class Doc2VecVocab(Word2VecVocab):
     def __init__(self, max_vocab_size=None, min_count=5, sample=1e-3, sorted_vocab=True, null_word=0):
         super(Doc2VecVocab, self).__init__(
             max_vocab_size=max_vocab_size, min_count=min_count, sample=sample,
             sorted_vocab=sorted_vocab, null_word=null_word)
 
-    def scan_vocab(self, documents, docvecs, progress_per=10000, trim_rule=None):
-        logger.info("collecting all words and their counts")
+    def _scan_vocab_multistream(self, input_streams, docvecs, workers, trim_rule):
+        manager = multiprocessing.Manager()
+        progress_queue = manager.Queue()
+
+        logger.info("Scanning vocab in %i processes.", min(workers, len(input_streams)))
+        pool = multiprocessing.Pool(processes=min(workers, len(input_streams)))
+
+        results = [
+            pool.apply_async(_scan_vocab_worker,
+                             (stream, progress_queue, self.max_vocab_size, trim_rule)
+                             ) for stream in input_streams
+        ]
+        pool.close()
+
+        unfinished_tasks = len(results)
+        total_words = 0
+        document_no = -1
+        while unfinished_tasks > 0:
+            report = progress_queue.get()
+            if report is None:
+                unfinished_tasks -= 1
+                logger.info("scan vocab task finished, processed %i documents and %i words;"
+                            " awaiting finish of %i more tasks", document_no + 1, total_words, unfinished_tasks)
+            elif isinstance(report, string_types):
+                logger.warning(report)
+            else:
+                num_words, num_documents = report
+                total_words += num_words
+                document_no += num_documents
+
+        corpus_count = document_no + 1
+        results = [res.get() for res in results]  # pairs (vocab, doclen2tags)
+        self.raw_vocab = reduce(utils.merge_dicts, [r[0] for r in results])
```

Comment:
`from functools import reduce`, Guido doesn't like `reduce` function, for this reason `reduce` was hidden in python3.

Reply:
done